spectator-api/src/main/java/com/netflix/spectator/api/patterns/PolledGauge.java

```diff
@@ -0,0 +1,402 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spectator.api.patterns;
+
+import com.netflix.spectator.api.Gauge;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Measurement;
+import com.netflix.spectator.api.Meter;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.Utils;
+
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * Helper for configuring a gauge that will receive a value by regularly polling the
+ * source in the background.
+ *
+ * <p>Example usage:</p>
+ *
+ * <pre>
+ *   Registry registry = ...
+ *   AtomicLong connections = PolledGauge.using(registry)
+ *     .withName("server.currentConnections")
+ *     .monitor(new AtomicLong());
+ *
+ *   // When a connection is added
+ *   connections.incrementAndGet();
+ *
+ *   // When a connection is removed
+ *   connections.decrementAndGet();
+ * </pre>
+ *
+ * <p>Polling frequency will depend on the underlying registry implementation, but users should
+ * assume it will be frequently checked and that the provided function is cheap. Users should
+ * keep in mind that polling will not capture all activity, just sample it at some frequency.
+ * For example, if monitoring a queue, then a gauge will only tell you the last sampled size
+ * when the value is reported. If more details are needed, then use an alternative type
+ * and ensure that all changes are reported when they occur.</p>
+ *
+ * <p>For example, consider tracking the number of currently established connections to a server.
+ * Using a polled gauge will show the last sampled number when reported. An alternative would
+ * be to report the number of connections to a {@link com.netflix.spectator.api.DistributionSummary}
+ * every time a connection is added or removed. The distribution summary would provide more
+ * accurate tracking such as max and average number of connections across an interval of time.
+ * The polled gauge would not provide that level of detail.</p>
+ *
+ * <p>If multiple values are monitored with the same id, then the values will be aggregated and
+ * the sum will be reported. For example, registering multiple gauges for active threads in
+ * a thread pool with the same id would produce a value that is the overall number
+ * of active threads. For other behaviors, manage it on the user side and avoid multiple
+ * registrations.</p>
+ */
+public final class PolledGauge {
```

Comment:
This class and `MonotonicCounter` have a lot of commonality resulting in significant code and documentation duplication. Consider distilling out the common bits.